### PR TITLE
[fix] Disallow wallet.export for signer backed wallets

### PR DESCRIPTION
**Pull request by jazz-cb**
### What changed? Why?

Disallow wallet.export for signer backed wallets

### CODE REVIEW

File: lib/coinbase/wallet.rb [hunk 1]
```diff
@@ -123,11 +123,15 @@ def wallets_api
...
     # @param client [Jimson::Client] (Optional) The JSON RPC client to use for interacting with the Network
     def initialize(model, seed: nil, address_models: [])
-      validate_seed_and_address_models(seed, address_models)
+      validate_seed_and_address_models(seed, address_models) unless Coinbase.use_server_signer?
 
       @model = model
-      @master = master_node(seed)
       @addresses = []
+
+      return if Coinbase.use_server_signer?
```
Comment (alex-stone) [inline lib/coinbase/wallet.rb:131]: We would still want to fetch the addresses in this case potentially?

Maybe we can lazily do that when calling `addresses`, but that's one thing I'd like to think through.

Currently calling `User#wallets` has a major N+1 in terms of fetching addresses. I'd like to potentially delay that fetch until we actually need to interact with the addresses?

Comment (jazz-cb) [reply lib/coinbase/wallet.rb]: @alex-stone as an interim fix, updated this logic to load addresses even while using signer.
Lets optimize the address fetching in the next release

File: lib/coinbase/wallet.rb [hunk 2]
```diff
@@ -245,6 +248,10 @@ def transfer(amount, asset_id, destination)
     # Exports the Wallet's data to a Data object.
     # @return [Data] The Wallet data
     def export
+      puts Coinbase.use_server_signer?
```
Comment (alex-stone) [inline lib/coinbase/wallet.rb:251]: Remove?

Comment (jazz-cb) [reply lib/coinbase/wallet.rb]: done

File: spec/unit/coinbase/wallet_spec.rb [hunk 3]
```diff
@@ -611,24 +611,45 @@
...
-    it 'allows for re-creation of a Wallet' do
-      wallet_data = seed_wallet.export
-      new_wallet = described_class
-                   .new(model, seed: wallet_data.seed, address_models: [address_model1, address_model2])
-      expect(new_wallet.addresses.length).to eq(2)
-      new_wallet.addresses.each_with_index do |address, i|
-        expect(address.id).to eq(seed_wallet.addresses[i].id)
+    context 'when using a server signer' do
+      let(:use_server_signer) { true }
+      let(:seed_wallet) do
```
Comment (alex-stone) [inline spec/unit/coinbase/wallet_spec.rb:645]:
```suggestion
      let(:wallet_without_seed) do
```

File: lib/coinbase/wallet.rb [hunk 4]
```diff
@@ -123,12 +123,15 @@ def wallets_api
...
-      @master = master_node(seed)
       @addresses = []
-      @private_key_index = 0
+
+      unless Coinbase.use_server_signer?
+        @master = master_node(seed)
+        @private_key_index = 0
+      end
 
       derive_addresses(address_models)
```
Comment (alex-stone) [inline lib/coinbase/wallet.rb:136]: Can we add an `.import` test with server signer true that we properly load the addresses?

File: spec/unit/coinbase/wallet_spec.rb (hunk 3 above)
Comment (jazz-cb) [reply spec/unit/coinbase/wallet_spec.rb]: fixed

File: lib/coinbase/wallet.rb (hunk 4 above)
Comment (jazz-cb) [reply lib/coinbase/wallet.rb:136]: done

**Review by John-peterson-coinbase — APPROVED** [review]

**Pull request merged**